games/applications/hidden_shift.py

```diff
@@ -0,0 +1,147 @@
+    secret_string = secret_string[::-1]    # reverse secret_string to fit qiskit'secret_string qubit ordering
+    for i_qubit in range(num_qubits):
+        if secret_string[i_qubit] == '0':   # skips every secret_string qubit == 0
+            continue
+        else:
+            the_shift.x(i_qubit)    # applies x gate to secret_string qubit == 1, shifts the qubits
+    the_shift.name = 'Shift'
+    return the_shift
+
+# Display circuit
+the_shift(num_qubits, secret_string).draw()     # display the_shift
+
+#------------------------------------------------------------
+
+def g_oracle(num_qubits):
+    g_oracle = QuantumCircuit(num_qubits)
+
+    # shifts the qubits
+    g_oracle.append(the_shift(num_qubits, secret_string), range(num_qubits))
+    
+    # apply the oracle U_f = (-1)^f(x) = F(x)
+    # x·y corresponds to controlled-z gates between every qubit and its second qubit
+    for i_qubit in range(int(num_qubits/2)):
+        g_oracle.cz(i_qubit, i_qubit+int(num_qubits/2))
+    
+    # h(x) = x·e_2 corresponds to the z gate applied to middle qubit
+    g_oracle.z(int(num_qubits/2)-1) 
+    
+    # undo the shift
+    g_oracle.append(the_shift(num_qubits, secret_string), range(num_qubits))
+    
+    # naming oracle
+    g_oracle.name = 'Oracle g'
+    return g_oracle
+
+# display g_oracle
+g_oracle(num_qubits).draw()
+
+#------------------------------------------------------------
+
+def f_oracle(num_qubits):
+    # oracle circuit that encodes Fourier Transform 
+    f_oracle = QuantumCircuit(num_qubits)
+    
+    # x·y corresponds to controlled-z gates between every qubit and its second qubit
+    for i_qubit in range(int(num_qubits/2)):
+        f_oracle.cz(i_qubit, i_qubit+int(num_qubits/2))
+    
+    # apply h(y) = y·e_2, corresponds to the z gate of the second qubit of the last register, aka the last qubit of the circuit.
+    f_oracle.z(int(num_qubits-1)) 
+    
+    # name oracle
+    f_oracle.name = 'Oracle f'
+    return f_oracle
+
+# display f_oracle
+f_oracle(num_qubits).draw()
+
+#------------------------------------------------------------
+
+hs_circuit = QuantumCircuit(num_qubits, num_qubits)
+
+# apply Hadamard gates to all qubits to make superposition
+for i_qubit in range(num_qubits):
+    hs_circuit.h(i_qubit)
+hs_circuit.barrier()    # barrier for visual separation
+
+# query oracle_g  
+hs_circuit.append(g_oracle(num_qubits), range(num_qubits))
+hs_circuit.barrier()
+
+# apply Hadamard gates to all qubits to perform a Boolean Fourier Transform
+for i_qubit in range(num_qubits):
+    hs_circuit.h(i_qubit)
+hs_circuit.barrier()
+
+# query oracle_f
+hs_circuit.append(f_oracle(num_qubits), range(num_qubits))
+hs_circuit.barrier()
+
+# retrieve the secret_string by performing the Boolean Fourier transform again aka apply h gates again
+for i_qubit in range(num_qubits):
+    hs_circuit.h(i_qubit)
+hs_circuit.barrier()
+
+# measure all qubits
+for i_qubit in range(num_qubits):
+    hs_circuit.measure(i_qubit, i_qubit)
```

Comment:
Please turn hs_circuit into a function that returns a QuantumCircuit instance. This is why line 147 is not working. I will leave a suggestion for you.
```suggestion

def hs_circuit(secret_string, num_qubits):
   hs_circuit = QuantumCircuit(num_qubits, num_qubits):

   # apply Hadamard gates to all qubits to make superposition
   for i_qubit in range(num_qubits):
       hs_circuit.h(i_qubit)
   hs_circuit.barrier()    # barrier for visual separation

   # query oracle_g  
   hs_circuit.append(g_oracle(num_qubits), range(num_qubits))
   hs_circuit.barrier()

   # apply Hadamard gates to all qubits to perform a Boolean Fourier Transform
   for i_qubit in range(num_qubits):
       hs_circuit.h(i_qubit)
   hs_circuit.barrier()
   
   # query oracle_f
   hs_circuit.append(f_oracle(num_qubits), range(num_qubits))
   hs_circuit.barrier()
   
   # retrieve the secret_string by performing the Boolean Fourier transform again aka apply h gates again
   for i_qubit in range(num_qubits):
       hs_circuit.h(i_qubit)
   hs_circuit.barrier()
   
   # measure all qubits
   for i_qubit in range(num_qubits):
       hs_circuit.measure(i_qubit, i_qubit)
       
   return hs_circuit
```

Reply:
followed your suggestion, I also made the function name and QuantumCircuit name different for linting purposes.